### PR TITLE
[luci] Introduce Context to QuantizeDequantizeWeights

### PR DESCRIPTION
**Pull request by jinevening**
This introduces Context struct to QuantizedDequantizeWeights.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/issues/8399
Draft PR: https://github.com/Samsung/ONE/pull/8460

### CODE REVIEW

**Comment by hyunsik-yoon** [comment]
I am not sure about the context about this change since draft (https://github.com/Samsung/ONE/pull/8460) does not seem to have this change. Could you explain a bit about the reason of this change?

**Comment by jinevening** [comment]
@hyunsik-yoon It was my mistake. I forgot to update the draft PR. I'll update it.

> Could you explain a bit about the reason of this change?

We're going to support mixed-precision, and the precision information of layers should be passed to both `QuantizeDequantizeWeightsPass` (fake quantization) and `QuantizeWithMinMaxPass` (quantization), because their behaviors are coupled.

For example, if we fake quantize with u8, then we should quantize with u8 (no s16 fake quant -> u8 quant).

File: compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h [hunk 1]
```diff
@@ -31,23 +31,38 @@ namespace luci
  */
 class QuantizeDequantizeWeightsPass : public logo::Pass
 {
+public:
+  struct Context
+  {
+    loco::DataType input_model_dtype = loco::DataType::Unknown;
+    loco::DataType output_model_dtype = loco::DataType::Unknown;
+    QuantizationGranularity granularity = QuantizationGranularity::ChannelWise;
```
Comment (jinevening) [inline compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:39]: I've initialized variables to avoid #8468.

Comment (hyunsik-yoon) [inline compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:39]: Can we create `QuantizationGranularity::Unknown` and initialized `granularity` with `unknown` just like `input_model_dtype` and `output_model_dtype`? 
Initializing to channel by dfault is good to some NPU but may not be good to some other mechanisms which ONE needs to support (in future, which I don't know now 😸).

Comment (jinevening) [reply compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:39]: I totally agree with you. I've posted  #8472 for this issue :)

File: compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h [hunk 2]
```diff
@@ -31,23 +31,38 @@ namespace luci
...
-                                                                                        granularity}
   {
-    // DO NOTHING
+    _ctx = std::make_unique<Context>();
+    {
+      _ctx->input_model_dtype = input_model_dtype;
+      _ctx->output_model_dtype = output_model_dtype;
+      _ctx->granularity = granularity;
+    }
   }
```
Comment (mhs4670go) [inline compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:58]: How about use initializer list?
```cxx
QuantizeDequantizeWeightsPass(loco::DataType input_model_dtype, loco::DataType output_model_dtype,
                                QuantizationGranularity granularity) : _ctx{std::make_unique<Context>(Context{input_model_dtype, output_model_dtype, granularity})}
{
  // DO NOTHING
}
```
Or
```cxx
// need constructor of `Context`
QuantizeDequantizeWeightsPass(loco::DataType input_model_dtype, loco::DataType output_model_dtype,
                                QuantizationGranularity granularity) : _ctx{std::make_unique<Context>(input_model_dtype, output_model_dtype, granularity)}
{
  // DO NOTHING
}
```

Comment (seanshpark) [reply compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:58]: I like @jinevening 's code as it's easy to read :)

**Review by hyunsik-yoon — APPROVED** [review]
LGTM, Thanks.

File: compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h (hunk 2 above)
Comment (jinevening) [reply compiler/luci/pass/include/luci/Pass/QuantizeDequantizeWeightsPass.h:58]: In addition to @seanshpark 's comment, I'll add a new variable to `Context`, which may change the constructor signature. So, I'd like to keep the current style :)

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM

**Pull request merged**